dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Threading;
+using Microsoft.SemanticKernel.Data;
+using Microsoft.SemanticKernel.Http;
+
+namespace Microsoft.SemanticKernel.Connectors.Weaviate;
+
+/// <summary>
+/// Class for accessing the list of collections in a Weaviate vector store.
+/// </summary>
+/// <remarks>
+/// This class can be used with collections of any schema type, but requires you to provide schema information when getting a collection.
+/// </remarks>
+public sealed class WeaviateVectorStore : IVectorStore
+{
+    /// <summary><see cref="HttpClient"/> that is used to interact with Weaviate API.</summary>
+    private readonly HttpClient _httpClient;
+
+    /// <summary>Optional configuration options for this class.</summary>
+    private readonly WeaviateVectorStoreOptions _options;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="WeaviateVectorStore"/> class.
+    /// </summary>
+    /// <param name="httpClient"><see cref="HttpClient"/> that is used to interact with Weaviate API.</param>
+    /// <param name="options">Optional configuration options for this class.</param>
+    public WeaviateVectorStore(HttpClient httpClient, WeaviateVectorStoreOptions? options = null)
+    {
+        Verify.NotNull(httpClient);
+
+        this._httpClient = httpClient;
+        this._options = options ?? new();
+    }
+
+    /// <inheritdoc />
+    public IVectorStoreRecordCollection<TKey, TRecord> GetCollection<TKey, TRecord>(string name, VectorStoreRecordDefinition? vectorStoreRecordDefinition = null)
+        where TKey : notnull
+        where TRecord : class
+    {
+        if (typeof(TKey) != typeof(Guid))
+        {
+            throw new NotSupportedException($"Only {nameof(Guid)} key is supported.");
+        }
+
+        if (this._options.VectorStoreCollectionFactory is not null)
+        {
+            return this._options.VectorStoreCollectionFactory.CreateVectorStoreRecordCollection<TKey, TRecord>(
+                this._httpClient,
+                name,
+                vectorStoreRecordDefinition);
+        }
+
+        var recordCollection = new WeaviateVectorStoreRecordCollection<TRecord>(
+            this._httpClient,
+            name,
+            new()
+            {
+                VectorStoreRecordDefinition = vectorStoreRecordDefinition
+            }) as IVectorStoreRecordCollection<TKey, TRecord>;
+
+        return recordCollection!;
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<string> ListCollectionNamesAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        using var request = new WeaviateGetCollectionsRequest().Build();
+
+        var response = await this._httpClient.SendWithSuccessCheckAsync(request, cancellationToken).ConfigureAwait(false);
```

Comment:
Should there be support for passing an API key with this request?

Reply:
Added a possibility to specify API key in `options` and added XML documentation for `HttpClient` that it can also be included via `DefaultRequestHeaders` property. Also added tests to verify it is used correctly.